src/dbnode/persist/fs/persist_manager.go

```diff
@@ -654,3 +657,70 @@ func (pm *persistManager) SetRuntimeOptions(value runtime.Options) {
 	pm.currRateLimitOpts = value.PersistRateLimitOptions()
 	pm.Unlock()
 }
+
+var (
+	indexVolumeIndexClaimsLock sync.Mutex
+	indexVolumeIndexClaims     = make(map[string]indexVolumeIndexClaim)
+)
+
+type indexVolumeIndexClaim struct {
+	volumeIndex int
+	blockStart  xtime.UnixNano
+}
+
+// claimNextIndexFileSetVolumeIndex ensures we never use the same volume index
+// no matter what for an index volume from the same process, we always just
+// select the next integer regardless if previous attempt to write to prior
+// index failed or succeeded. This is ok since volume indexes do not need
+// to be contiguous, they just need to be ascending.
+func claimNextIndexFileSetVolumeIndex(
+	opts Options,
+	namespaceMetadata namespace.Metadata,
+	blockStartTime time.Time,
+) (int, error) {
+	indexVolumeIndexClaimsLock.Lock()
+	defer indexVolumeIndexClaimsLock.Unlock()
+
+	filePathPrefix := opts.FilePathPrefix()
+	nowFn := opts.ClockOptions().NowFn()
+
+	// Reap any old entries that are out of retention.
+	retOpts := namespaceMetadata.Options().RetentionOptions()
+	indexOpts := namespaceMetadata.Options().IndexOptions()
+	rp, bs, t := retOpts.RetentionPeriod(), indexOpts.BlockSize(), nowFn()
+	earliestBlockStart := retention.FlushTimeStartForRetentionPeriod(rp, bs, t)
+	earliestBlockStartUnixNanos := xtime.ToUnixNano(earliestBlockStart)
+	for key, claim := range indexVolumeIndexClaims {
```

Comment:
nit: prob a hyper optimization but would it be better to just iterate over block starts beginning w/ edge of retention and going backwards in time until we can't find any claim and bail then? That way we're only ever iterating over block starts that are out of retention (if any). If there are none, it will just check the map once and bail early.